packages/cli/lib/tab-completion.js

```diff
@@ -0,0 +1,224 @@
+// Copyright IBM Corp. 2018,2020. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+'use strict';
+
+const debug = require('./debug')();
+const tabtab = require('tabtab');
+
+const tabCompletionCommands = [
+  'install-completion',
+  'uninstall-completion',
+  'completion',
+];
+
+const commands = [
```

Comment:
I think it's possible to build these lists from generator metadata without hard-coding them. Let's look at how `help` is implemented. `--help` lists available commands and their options.

Reply:
@raymondfeng Thank you for your feedback. I've checked the current solution for getting all possible commands. And I was thinking about it before to take `env.getGeneratorsMeta()` solution. But I see a problem with this solution. Even if I get commands from `env.getGeneratorsMeta()`, we don't have the possibility to take command's options from yeoman. Therefore hardcoded options and not-hardcoded commands should have been supported that is not good from my perspective.

But I investigated `yeoman-environment` and found internal `_options` property. What do you think about this proposal: add `getOptions` method inside `BaseGenerator` class.

```
...
getOptions() {
  return this._options;
}
...
```

this._options view for controller:

```
{ help:
   { name: 'help',
     type: [Function: Boolean],
     alias: 'h',
     description: 'Print the generator\'s options and usage' },
  'skip-cache':
   { name: 'skip-cache',
     type: [Function: Boolean],
     description: 'Do not remember prompt answers',
     default: false },
  'skip-install':
   { name: 'skip-install',
     type: [Function: Boolean],
     description: 'Do not automatically install dependencies',
     default: false },
  'force-install':
   { name: 'force-install',
     type: [Function: Boolean],
     description: 'Fail on install dependencies error',
     default: false },
  'ask-answered':
   { type: [Function: Boolean],
     description: 'Show prompts for already configured options',
     default: false,
     name: 'ask-answered',
     hide: false },
  controllerType:
   { type: [Function: String],
     required: false,
     description: 'Type for the controller',
     name: 'controllerType',
     hide: false },
  config:
   { type: [Function: String],
     alias: 'c',
     description: 'JSON file name or value to configure options',
     name: 'config',
     hide: false },
  yes:
   { type: [Function: Boolean],
     alias: 'y',
     description:
      'Skip all confirmation prompts with default or provided value',
     name: 'yes',
     hide: false },
  format:
   { type: [Function: Boolean],
     description: 'Format generated code using npm run lint:fix',
     name: 'format',
     hide: false } }

```

After getting options, I can filter and convert them to a fit view. It makes possible to use dynamic commands and options.

Reply:
I'm adding code to get options and args for each command.

Reply:
It's better solution) Thank you